nav2_bt_navigator/src/navigators/navigate_through_poses.cpp

```diff
@@ -116,16 +122,15 @@ NavigateThroughPosesNavigator::onLoop()
       feedback_utils_.global_frame, feedback_utils_.robot_frame,
       feedback_utils_.transform_tolerance))
   {
-    RCLCPP_ERROR(logger_, "Robot pose is not available.");
-    return;
+    throw nav2_core::NavigatorPoseNotAvailable("Robot pose is not available.");
   }
 
   try {
     // Get current path points
     nav_msgs::msg::Path current_path;
     if (!blackboard->get(path_blackboard_id_, current_path) || current_path.poses.size() == 0u) {
       // If no path set yet or not meaningful, can't compute ETA or dist remaining yet.
-      throw std::exception();
+      throw nav2_core::NavigatorInvalidPath("no valid path available");
```

Comment:
So this is setup so that if we run this feedback callback before the actual navigation system is all warmed up (since this is called on a regular timer which is independent of the navigation logic's current state execution). That way, if we don't have a path/goal/etc yet, we just skip providing meaningless feedback ETAs and distances based on a lack of data yet. We don't need this to be a coded exception that should be propogated anywhere. This exception is purely for conciseness of the code that stays internal to the function

Reply:
All the comments associated with `current_error_code_/current_error_msg_` and the introduction of specific exceptions relate to this point, so I will attempt to layout the reasoning behind it here.  

I am pushing back but will accept it, if you subsequently overrule.  I am aiming to provide ALL reasons that a navigator action fails to an action client, ie an action client should not have to view server log files to find out why their request failed.

First up I have a general built-in aversion to catch(...), except inside a main() where all hope is finally lost.  Using a naked std::exception() without even an hint of where it came from, ie with at least a string that *could* be used to hunt it down is worse.  I get that this particular case is internal to this function, but its a style with a foot gun IMHO.  To then do what I think is more correct I followed similar patterns used elsewhere in the code.  

The decision to do nothing about (not propogate) this particular exception/exit point, NavigatorInvalidPath, and elsewhere NavigatorPoseNotAvailable are now made in the catch block.  If those decisions are correct or not can now be manipulated on a case by case basis.  They can also have their own breakpoint in a debugger.  I have left comments // Ignore?? were I wasn't sure if they should be or not, BUT the current behavior is intended to be maintained(Need to double check).

Next, the use of `current_error_code_` and `current_error_msg_` that get defined in the base(intermediate) class `BehaviourTreeNavigator`. 

This my attempt to ALSO capture reasons for action failure within the setup of the navigator itself. Not just its behavior tree processing.  Note that I added a call to `populateErrorCode`, if `on_goal_received_callback_()` fails in `BtActionServer<ActionT>::executeCallback()` (see nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp)

`current_error_code_` and `current_error_msg_` are also intended to be picked up and populate the action `ActionT::Result::SharedPtr` result in `NavigateThroughPosesNavigator::goalCompleted()` and `NavigateToPoseNavigator::goalCompleted()`.  Previously the implementation of these functions were empty.
 
After following this through I can see where I have not completed the task.  I need to fix `BtActionServer<ActionT>::populateErrorCode()` to pick the "error_msg" out of the `blackboard_` and use `current_error_code_` and `current_error_msg_` if they are relevant.  I also need to clean `current_error_code_` and `current_error_msg_` in `BtActionServer<ActionT>::cleanErrorCodes()`

I also need to study your previous comment https://github.com/ros-navigation/navigation2/pull/4460#discussion_r1672732244

So yeah, overall, still in draft and needs testing.  Thanks for taking a look and prompting.

Reply:
> I am aiming to provide ALL reasons that a navigator action fails to an action client, ie an action client should not have to view server log files to find out why their request failed.

Understood and I agree, but I think there's been some misunderstanding in this case. The `onLoop` method is used to do nothing other than populate the `ActionT::Feedback` message for the users while a behavior tree is running. The `onLoop` is actually called as part of the BT execution loop, in which if this callback is being triggered, means that there is no failure in the action or anything to report. Exceptions here are all internal to the method for handling the lack of some information on the blackboard at the start of new actions. No errors in the BT are processed or caught here. 

> Using a naked std::exception() without even an hint of where it came from, ie with at least a string that could be used to hunt it down also doesn't help. 

It is so generically typed because its[ immediately caught here](https://github.com/ros-navigation/navigation2/blob/main/nav2_bt_navigator/src/navigators/navigate_through_poses.cpp#L167-L169) to ignore populating some part of the `ActionT::Feedback` message. This is not an action failure, its just skipping a chunk of code because there isn't a path yet to compute an ETA or distance remaining with respect to yet. 

I, however, don't disagree that there are improvements that could be made here, in terms of clarity and/or removing the exception and making all that block under a nested `if (blackboard->get(path_blackboard_id_, current_path) && current_path.poses.size() > 0u)`. 

> The decision to do nothing about (not propogate) this particular exception/exit point, NavigatorInvalidPath, and elsewhere NavigatorPoseNotAvailable are now made in the catch block.

I think I have communicated this above, but if not clear, please let me know. These are not error cases that have to do with the navigator or navigation happening in the BT. These are just checks if we have the data we require on the blackboard to populate some of the `ActionT::Feedback` fields. We use exceptions in Nav2 typically w.r.t. the error codes and error propagation, but this is one example of a use of exceptions that is not related to errors. It could be replaced with an if statement to ignore populating elements of the feedback which are not yet ready to be processed yet. 

Imagine you had a planner that took 15 seconds to plan for whatever reason. The first 15 * loop_rate times `onLoop` is called, there's no `path` on the blackboard to compute ETA / distances with respect to. So, we ignore those fields until we're in a navigating state and have a path to work with. That's why the catch for the generic `exception()` just said `// ignore`. 

With all that in mind, I think that should provide context that isn't necessary for the `onLoop` changes. If you have a counter opinion, I can respect that and happy to hear it, but I'm not sure your previous point has the correct understanding of what's happening in the method and where this fits into the system. The actual BT error code / messages are part of the `ActionT::Result` on the termination of the action which is in other parts of the code which you have identified in `bt_action_server_impl.hpp`'s `populateErrorCode`

---

> This my attempt to ALSO capture reasons for action failure within the setup of the navigator itself. Not just its behavior tree processing. Note that I added a call to populateErrorCode, if on_goal_received_callback_() fails in BtActionServer<ActionT>::executeCallback() (see nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp)

Ah yes, I do see that now. We still need to fix the actual population of `highest_priority_error_msg` for that to be given to the result `result->error_msg = highest_priority_error_msg;`, but point is taken here and that's a good addition with `FAILED_TO_LOAD_BEHAVIOR_TREE` and `TF_ERROR`! I don't argue with those, except to rename to `GOAL_TRANSFORMATION_ERROR` to be specific that the TF error was wrt transforming the goal to the global frame - but that's a nitpick.

I'm only suggesting reverting the `onLoop` stuff now, the things w.r.t. on goal reception do get propagated correctly and are useful indeed.

>  current_error_code_ and current_error_msg_ are also intended to be picked up and populate the action ActionT::Result::SharedPtr result in NavigateThroughPosesNavigator::goalCompleted() and NavigateToPoseNavigator::goalCompleted(). Previously the implementation of these functions were empty.

I don't think that these work properly? I'm not sure that `goalCompleted` does anything actually. You touched the right part of the code in `populateErrorCode` which that is done. We have a list of BT Navigator parameterized blackboard keys that have the error codes from each of the servers are reviewed and we use the highest priority error code of the bunch, if there are multiple. This is where we should add in an analog thing for the error messages and possibly change the way we parameterize the error code IDs so that we have the ID correlated to the message blackboard variable.